### PR TITLE
Layouts (stage 4): print layouts of type variables in a comment when the user asks twice

### PR DESCRIPTION
**Pull request by ncik-roberts**
As discussed internally. When the user queries a type expression twice, we print `(* 'a_1 : layout_1, ..., 'a_n : layout_n *)` for all type variables `'a_i` with non-boring layout `layout_i`.

Note: type variables are printed in the reverse order from what you'd expect. This is a general issue with type printing in flambda-backend; I'll post a PR there rather than fixing it in this PR.

### CODE REVIEW

File: tests/test-dirs/type-enclosing/jane-street.t/layouts.ml [hunk 1]
```diff
@@ -25,10 +25,20 @@ let poly_client2 x = poly2 x
 let poly_client3 x = poly3 x
 let poly_client4 x = poly4 x
```
Comment (alanechang) [inline tests/test-dirs/type-enclosing/jane-street.t/layouts.ml:26]: Can we show that non-constant sort variables are not printed? Something along the lines of `let f x = x` perhaps

**Review by alanechang — APPROVED** [review]
Looks good overall! Left some comments related to implementation details.

File: src/ocaml/typing/printtyp.ml [hunk 2]
```diff
@@ -3011,9 +3011,40 @@ let shorten_class_type_path env p =
...
+               List.iter
+                 (fprintf ppf ", %a" annotated_qtv)
+                 qtvs)
+
 let type_declaration_for_merlin = type_declaration
 
 (* Drop merlin-only arguments from exported interface *)
+
+let prepared_type_scheme x y : unit = prepared_type_scheme x y
+let type_scheme x y : unit = type_scheme x y
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply src/ocaml/typing/printtyp.ml:3047]: Yes, we can, thanks. This was an artifact of a merge conflict resolution.

File: src/ocaml/typing/printtyp.ml [hunk 3]
```diff
@@ -3011,9 +3011,40 @@ let shorten_class_type_path env p =
...
+      in
+      match qtvs with
+      | [] -> ()
+      | qtv :: qtvs ->
+          fprintf ppf " @[(* @[%a%t@] *)@]"
+            annotated_qtv qtv
+            (fun ppf ->
+               List.iter
+                 (fprintf ppf ", %a" annotated_qtv)
+                 qtvs)
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply src/ocaml/typing/printtyp.ml]: `Format.pp_print_list` is a great fit, thanks.

I'm not a fan of factoring out the code to an inner function, `let print_jkind_comment () = ...`. That forces you to remember, while scanning the code, whether you're in a "strict" context or not. (When does this code actually run?) I recognize that the logic is a bit complex, though. I've factored it out to another top-level function, hopefully with a more intelligible name.

File: tests/test-dirs/type-enclosing/jane-street.t/layouts.ml (hunk 1 above)
Comment (ncik-roberts) [reply tests/test-dirs/type-enclosing/jane-street.t/layouts.ml:26]: Sure, done.

**Pull request merged**